### PR TITLE
VLAYER

### PR DESCRIPTION
**Pull request by PHIDIAS0303**

### CODE REVIEW

**Comment by Cooldude2606** [comment]
This change can not be made because `surface.daytime` can be modified from other scripts.
For example if `surface.always_day` or `surface.freeze_daytime` are true then `time` will continue to tick when it should be constant.

**Comment by PHIDIAS0303** [comment]
but the main thing is that I need the tick to do the day-night calculation of solar power ...

**Comment by Cooldude2606** [comment]
> but the main thing is that I need the tick to do the day-night calculation of solar power ...

For it to be accurate for vlayer solar it must use `surface.daytime` because this is used to calculate solar power production not game tick.

File: modules/control/vlayer.lua [hunk 1]
```diff
@@ -440,12 +440,14 @@ function vlayer.get_statistics()
         total_surface_area = vlayer_data.properties.total_surface_area,
         used_surface_area = vlayer_data.properties.used_surface_area,
         remaining_surface_area = math.max(vlayer_data.properties.total_surface_area - vlayer_data.properties.used_surface_area, 0),
+        multiplier = get_production_multiplier(),
```
Comment (Cooldude2606) [inline modules/control/vlayer.lua:443]: `multiplier` is vauge, use full name `production_multiplier`

File: modules/control/vlayer.lua [hunk 2]
```diff
@@ -440,12 +440,14 @@ function vlayer.get_statistics()
...
         energy_production = vlayer_data.properties.production * mega * get_production_multiplier(),
         energy_sustained = vlayer_data.properties.production * mega * get_sustained_multiplier(),
         energy_capacity = vlayer_data.properties.capacity * mega,
         energy_storage = vlayer_data.storage.energy,
         day = math.floor(game.tick / vlayer_data.surface.ticks_per_day),
-        time = math.floor(vlayer_data.surface.daytime * vlayer_data.surface.ticks_per_day)
+        day_length = vlayer_data.surface.ticks_per_day,
+        time = math.floor(game.tick % vlayer_data.surface.ticks_per_day)
```
Comment (Cooldude2606) [inline modules/control/vlayer.lua:450]: Unable to accept the use of `game.tick` when it introduces inconsistencies. Add a `tick` property and combine with your `day_length` property in combinators if you absolutely must.

**Review by Cooldude2606 — CHANGES_REQUESTED** [review]

**Review by Cooldude2606 — CHANGES_REQUESTED** [review]
Not what I meant why I said add a tick property. I meant a `tick` signal that is equal to `game.tick` so that you can do `tick % day_length` in a combinator and return `time` to be based on `daytime`.

File: modules/gui/vlayer.lua [hunk 3]
```diff
@@ -10,8 +10,20 @@ local Event = require 'utils.event' --- @dep utils.event
 local format_number = require('util').format_number --- @dep util
 local config = require 'config.vlayer' --- @dep config.vlayer
 local vlayer = require 'modules.control.vlayer'
+local Selection = require 'modules.control.selection' --- @dep modules.control.selection
+local SelectionConvertArea = 'ConvertArea'
```
Comment (Cooldude2606) [inline modules/gui/vlayer.lua:14]: This must be a unique string. Currently it conflicts with waterfill.
```suggestion
local SelectionConvertArea = 'VlayerConvertChest'
```

File: modules/gui/vlayer.lua [hunk 4]
```diff
@@ -59,26 +71,67 @@ local function format_energy(amount, unit)
...
+    local disp = frame.container['vlayer_st_2'].disp.table
+    local target = vlayer_control_type_list[disp[vlayer_gui_control_type.name].selected_index]
+
+    if #entities == 0 then
         player.print{'vlayer.steel-chest-detect'}
-        return
+        return nil
+
+    elseif #entities > 1 then
+        player.print{'vlayer.result-unable', {'vlayer.control-type-' .. target:gsub('_', '-')}}
```
Comment (Cooldude2606) [inline modules/gui/vlayer.lua:93]: Can you add a reason to this, eg `Unable to build Energy interface: Multiple steel chests selected`
Also add a reason to the other places this is used. ie `At interface limit` and `Not enough room`

**Review by Cooldude2606 — CHANGES_REQUESTED** [review]
I really like the change to use the selection tool. It all works correctly but the error messages need improving.

**Review by Cooldude2606 — APPROVED** [review]

**Pull request merged**